File: lib/merge.ts
export function mergeText(input: string, data: Record<string, unknown>) {
  return input.replace(/{{\s*([^{}]+?)\s*}}/g, (_, key) => String(data[key] ?? ''));
}
export function escapeHtml(s: string) {
  return s.replace(/[&<>"']/g, c => ({'&':'&amp;','<':'&lt;','>':'&gt;','"':'&quot;',"'":'&#039;'}[c] as string));
}
export function textToHtml(s: string) { return escapeHtml(s).replace(/\n/g, '<br>'); }
export function spreadsheetIdFrom(input: string) {
  const m = input.match(/\/spreadsheets\/d\/([a-zA-Z0-9-_]+)/);
  return m?.[1] ?? input.trim();
}

File: backend/app/services/merge.py
import html
import re

TOKEN = re.compile(r"{{\s*([^{}]+?)\s*}}")


def merge_text(value: str, data: dict) -> str:
    normalized = {str(key).strip().lower(): val for key, val in data.items()}
    return TOKEN.sub(lambda match: str(normalized.get(match.group(1).strip().lower(), "")), value)


def text_to_html(value: str) -> str:
    return html.escape(value).replace("\n", "<br>")


def spreadsheet_id(value: str) -> str:
    match = re.search(r"/spreadsheets/d/([a-zA-Z0-9-_]+)", value)
    return match.group(1) if match else value.strip()

